In [2]:
import numpy as np
import pandas as pd
import io
import os
from dags.airc.modules.utlis import _get_minio_object, _minio_object_to_dataframe

In [3]:
BUCKET_NAME = os.environ.get("BUCKET_NAME")
DATA_PATH = os.environ.get("DATA_PATH")

In [4]:
minio_object = _get_minio_object(BUCKET_NAME, "data/canada_house.csv")
df = _minio_object_to_dataframe(minio_object)

# df = pd.read_csv("/home/aircsrv5/Quan/DataOps/DataOps-bootstrap/demo/include/data/full_canada.csv", index_col=False)

/tmp/ipykernel_2681007/3962594733.py:4: DtypeWarning: Columns (42,87,93,259) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/home/aircsrv5/Quan/DataOps/DataOps-bootstrap/demo/include/data/full_canada.csv", index_col=False)


The dataset is quite big. Let's drop null rows and columms

In [5]:
# Drop duplicated rows
df = df.drop_duplicates()

# Drop duplicate columns
df = df.T.drop_duplicates().T
df.shape

(98649, 32)

With option pd.set_option('display.max_columns', None) we can see all features

In [6]:
pd.set_option('display.max_columns', None)
df.head()

streetAddress addressLocality  \
0                                                NaN             NaN   
3                                                NaN             NaN   
5                                                NaN             NaN   
6                                                NaN             NaN   
7  1701 Coursier Avenue Unit# 1301, Revelstoke, B...      Revelstoke   

  addressRegion postalCode   latitude   longitude  \
0           NaN        NaN        NaN         NaN   
3           NaN        NaN        NaN         NaN   
5           NaN        NaN        NaN         NaN   
6           NaN        NaN        NaN         NaN   
7            BC     V0E2S3  50.976585 -118.173149   

                                         description     price priceCurrency  \
0                                                NaN       NaN           NaN   
3                                                NaN       NaN           NaN   
5                                                NaN       NaN           NaN   
6                                                NaN       NaN           NaN   
7  Don't miss out on this rare 3 bedroom, 2 bath ...  839000.0           CAD   

  property-beds property-baths property-sqft Acreage Air Conditioning  \
0           0.0            NaN           NaN     NaN              NaN   
3           NaN            0.0           NaN     NaN              NaN   
5           2.0            2.0           NaN     NaN              NaN   
6           4.0            3.0           NaN     NaN              NaN   
7           3.0            2.0           891     NaN              NaN   

  Basement Bath Exterior                    Features Fireplace Garage  \
0      NaN  NaN      NaN                         NaN       NaN    NaN   
3      NaN  NaN      NaN                         NaN       NaN    NaN   
5      NaN  NaN      NaN                         NaN       NaN    NaN   
6      NaN  NaN      NaN                         NaN       NaN    NaN   
7      NaN  0.0      NaN  Underground, Heated Garage       NaN    NaN   

     Heating    MLS® # Property Tax Property Type Roof  \
0        NaN       NaN          NaN           NaN  NaN   
3        NaN       NaN          NaN           NaN  NaN   
5        NaN       NaN          NaN           NaN  NaN   
6        NaN       NaN          NaN           NaN  NaN   
7  Heat Pump  10334677       $3,478         Condo  NaN   

                     Sewer Square Footage Subdivision Waterfront  \
0                      NaN            NaN         NaN        NaN   
3                      NaN            NaN         NaN        NaN   
5                      NaN            NaN         NaN        NaN   
6                      NaN            NaN         NaN        NaN   
7  Municipal sewage system       891 SQFT         NaN        NaN   

                      Parking               Flooring Fireplace Features  
0                         NaN                    NaN                NaN  
3                         NaN                    NaN                NaN  
5                         NaN                    NaN                NaN  
6                         NaN                    NaN                NaN  
7  Underground, Heated Garage  Tile, Carpeted, Vinyl                NaN

There is no need in data without address or price, so we can drop all rows with null values in "streetAddress", "addressLocality", "addressRegion", "price" columns:

In [7]:
df = df.dropna(subset=["streetAddress", "addressLocality", "addressRegion", "price"])

In [8]:
df.shape

(97811, 32)

# Square Footage and property-sqft

Columns 'Square Footage', 'property-sqft' have the same information. Let's combine them in one column. 

In [9]:
df[['Square Footage', 'property-sqft']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 97811 entries, 7 to 133724
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Square Footage  46732 non-null  object
 1   property-sqft   46489 non-null  object
dtypes: object(2)
memory usage: 2.2+ MB


In [10]:
df.loc[:, 'Square Footage new'] = df['Square Footage'].str.replace(' SQFT', '', regex=False)
df.loc[:, 'Square Footage new'] = df['Square Footage new'].fillna(df['property-sqft'])

df[['property-sqft', 'Square Footage', 'Square Footage new']].head()

property-sqft Square Footage Square Footage new
7            891       891 SQFT                891
8          1,881     1,881 SQFT              1,881
9            912       912 SQFT                912
10         4,374     4,374 SQFT              4,374
11           NaN            NaN                NaN

In [11]:
df[['Square Footage new']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 97811 entries, 7 to 133724
Data columns (total 1 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Square Footage new  46732 non-null  object
dtypes: object(1)
memory usage: 1.5+ MB


# Acreage

In [12]:
df.Acreage.head()

7      NaN
8     0.32
9      NaN
10    0.69
11    0.14
Name: Acreage, dtype: object

"Acreage" typically refers to a large piece of land, usually measured in acres. In real estate, it implies that the property includes a significant amount of land, often used for farming, ranching, or simply as a large private estate. So, Nan values can be replaced to 0

In [13]:
df["Acreage"] = df["Acreage"].fillna(0)
df.Acreage.head()

7     0.00
8     0.32
9     0.00
10    0.69
11    0.14
Name: Acreage, dtype: float64

# property-bath

In [14]:
df[['property-baths', 'Bath']].describe()

property-baths     Bath
count          97041.0  26451.0
unique            26.0      9.0
top                2.0      1.0
freq           28083.0  17532.0

In [15]:
# Get unique values from both columns
unique_values = pd.Index(df['property-baths'].dropna().unique()).union(df['Bath'].dropna().unique())

# Create a DataFrame with unique values as index
counts_df = pd.DataFrame(index=unique_values, columns=['property-baths', 'Bath'])

# Count occurrences and fill the DataFrame
counts_df['property-baths'] = df['property-baths'].value_counts().reindex(unique_values, fill_value=0)
counts_df['Bath'] = df['Bath'].value_counts().reindex(unique_values, fill_value=0)

print(counts_df)

       property-baths   Bath
0.0             14428   7630
1.0             16867  17532
2.0             28083   1212
3.0             20532     59
4.0             11252     14
5.0              3020      1
6.0              1469      0
7.0               654      0
8.0               359      0
9.0               186      1
10.0               88      0
11.0               45      0
12.0               31      0
13.0                7      0
14.0                4      0
15.0                3      0
16.0                1      0
17.0                1      0
18.0                2      0
19.0                1      0
21.0                2      1
23.0                1      0
25.0                1      0
26.0                2      0
42.0                1      0
802.0               0      1
803.0               1      0


In [16]:
pd.set_option('display.max_colwidth', None)

df[df['property-baths'] == 42.0][['property-baths', 'Bath', 'description', 'price']]

property-baths Bath  \
82880           42.0  NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 description  \
82880  Discover this remarkable 24,000 sq ft commercial-residential mixed-use property offering incredible income potential and a proven high return on investment. Situated in a prime location, this asset is a standout opportunity for seasoned investors and developers alike.Key Features:1.Outstanding Income Return:oCurrent net profit: $675,557.13 annually, equating to a 12.75% net income return.oPost-conversion potential: $766,182 annual net profit, yielding an enhanced 14.45% net return.2.High-Value Asset with Significant Upside:oAppraised at $7.25M by a bank-approved appraiser.oMarket value projected to exceed $9M upon conversion to a 100% residential property.3.Existing Configuration:o12,000 sq ft of prime retail and office space.oResidential units: 4 studios, 8 one-bedroom apartments, and 8 two-bedroom apartments.4.Conversion Potential:oAfter transformation, the property will feature 6 studios, 8 one-bedroom apartments, and 28 two-bedroom apartments, catering to high-demand residential needs.5.Renovation Incentives:oGuaranteed renovation cost: Only $650,000, after factoring in city-supported CIP grants and interest-free loans.oBoosted rental income: Conversion results in a $90,625 annual increase in net rental income.6.Favorable Financing Options:oCMHC loan eligibility upon issuance of building permits.oLoan covers 100% of conversion costs at a competitive 4% interest rate.Why This Property?With strong current returns, significant upside from planned conversions, and the city-backed incentives to minimize costs, this property represents a rare blend of high yields, manageable risk, and long-term value appreciation.Don't miss this exceptional investment opportunity in the thriving City of Sudbury. (id:38686)   

           price  
82880  5388888.0

In [17]:
df[df['property-baths'] == 803][['property-baths', 'Bath', 'description', 'price']]

property-baths   Bath  \
69009          803.0  802.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     description  \
69009  This 2 bedroom home is located on Blackhead Road and looking for new owners. The main floor has an eat in kitchen, 2 bedrooms, living room, 4 piece bathroom and a porch. There is a door located off the kitchen to a large patio that would be great for those summer BBQ's. There is a room developed in the basement that would make a great office or a third bedroom for the teenager who need or want their own space. There is also a large rooms that would make an ideal rec room when it is completed. Lots of room in the back garden and an area that is fenced and great for the animals to run and play. There is also a pad already poured if someone is looking to build an outbuilding. This home required some work like painting, flooring and any upgrades you may want to do but there is no doubt that this home has lots of potential. (id:38686)   

          price  
69009  198000.0

In [18]:
df.loc[df['property-baths'] == 803, 'property-baths'] = 1

In [19]:
df[['property-baths']].value_counts()

property-baths
2.0               28083
3.0               20532
1.0               16868
0.0               14428
4.0               11252
5.0                3020
6.0                1469
7.0                 654
8.0                 359
9.0                 186
10.0                 88
11.0                 45
12.0                 31
13.0                  7
14.0                  4
15.0                  3
18.0                  2
21.0                  2
26.0                  2
16.0                  1
17.0                  1
19.0                  1
23.0                  1
25.0                  1
42.0                  1
Name: count, dtype: int64

# Features

In [20]:
df.Features.head()

7                          Underground, Heated Garage
8     Attached Garage, RV, See Remarks, Heated Garage
9                                            Aluminum
10                                    Attached Garage
11                                          Other, RV
Name: Features, dtype: object

The column 'Features' contains detailed information about a house, including parking, garage, pool, heating, garden, and other. Some rows may have multiple features listed, while others might only contain a few or even none. Additionally, the 'Garage' column can sometimes be empty, even when garage-related details are present in the 'Features' column. This makes it crucial to extract and analyze all values from 'Features' to ensure no important information is missed.

In [21]:
df['Features'].unique().shape

(5338,)

We have 5,338 different values in the 'Feature' column. The problem is that each cell can contain a list of information, with potential relationships between features in the dataset. We can define some keywords for each feature, extract relevant values, and fill in the gaps.

In [22]:
##### Extract unique features
def get_unique_values(column):
    
    unique_features = set()
    
    for features in df[column].dropna():
        # Remove brackets and extra quotes
        cleaned_features = features.replace("[", "").replace("]", "").replace('"', '').replace("'", "")
        # Split by comma
        feature_list = [f.strip() for f in cleaned_features.split(",")]
        unique_features.update(feature_list)

    return unique_features

In [23]:
unique_features = get_unique_values('Features')
print(len(unique_features), unique_features)

268 {'Oversized', 'Stone', 'Interlocking Driveway', '', 'Insul Brick', 'Gas', 'Cul-De-Sac', 'Plug-In', 'Secured', 'Partially Fenced', 'Titled', 'Carport Double', 'Electric Gate', 'Log', 'Tandem Parking', 'View City', 'Hardboard', 'Front Drive Access', 'Rock Imitation', '3 Car Attached', 'Double Indoor', 'Wood siding', 'Direct vent', 'Public Swimming Pool', 'Wood shingles', 'Breezeway', 'No Parking', 'Single Carport', 'Integrated Garage', 'Outdoor Grill', 'Private Electric Vehicle Charging Station(s)', 'Garden', 'Underground', 'Double Garage Attached', 'Sloping Lot', 'River Valley View', 'Waterfront Property', 'Garage Quad+', 'Underground Parking', 'Insert', 'Workshop in Garage', 'Unknown', 'Concrete Driveway', 'In Garage Electric Vehicle Charging Station(s)', 'Gated Community', 'RV Parking Avail.', 'Carport; Single', 'Brick Veneer', 'Insulated', 'Unpaved', 'Side By Side', 'Shared Driveway', 'Interlocked', 'Private Park Access', 'Parking Pad Cement/Paved', 'Detached Garage', 'Flat Site'

Now we got 268 different unique values. Let's fill the gap in Garage column

## Garage

In [24]:
df.Garage.value_counts()

Garage
Yes    9602
Name: count, dtype: int64

Column 'Garage' contains only 'Yes' or 'No' values. The information about type of garage, how namy of them in the house could contain columns 'Feature' or 'Parking'. Let's combine them.

In [25]:
unique_parking_features = get_unique_values('Parking') | unique_features
print(len(unique_parking_features), unique_parking_features)

268 {'', 'Gas', 'Log', 'Titled', 'Front Drive Access', 'Wood siding', 'Public Swimming Pool', 'Wood shingles', 'Integrated Garage', 'Outdoor Grill', 'Private Electric Vehicle Charging Station(s)', 'Garden', 'Underground', 'Waterfront Property', 'Underground Parking', 'Insert', 'Workshop in Garage', 'In Garage Electric Vehicle Charging Station(s)', 'Concrete Driveway', 'Unknown', 'Gated Community', 'Carport; Single', 'Brick Veneer', 'Insulated', 'Unpaved', 'Interlocked', 'Detached Garage', 'Flat Site', 'Rear Drive Access', 'Private Yard', 'Golf Nearby', 'Boating', 'View Downtown', 'Low Maintenance Yard', 'Offset', 'Garage Faces Rear', 'River View', 'Carport', 'Single Garage Attached', 'Rear', 'Stall', 'Siding', '6 Car Detached', 'No Garage', 'Surfaced', 'Off Street', 'Other', 'Garage/Laneway Suite', 'Awning(s)', 'Triple Garage Detached', 'Vinyl siding', 'Gravel', 'Cedar shingles', 'Aluminum/Vinyl', 'Parking Space(s)', 'Boat', 'Aggregate', 'Fenced', '2 Car Detached', 'No Back Lane', 'Cre

Let's select from this list only garage related terminology:

In [26]:
garage_keywords = {"garage", "carport", "grge", "attached", "detached", "tandem", "car"}

# The selected keywords cover different terms related to garages and carports:

#     "garage" → Captures any mention of a garage.
#     "carport" → Includes carports, which serve a similar function.
#     "grge" → A common abbreviation for "garage" (e.g., "DetachedGrge/Carport").
#     "attached" → Covers garages that are attached to the house.
#     "detached" → Covers garages that are separate from the house.
#     "tandem" → Represents tandem garages, where cars are parked one behind the other.
#     "car" → Captures features mentioning a car (e.g., "2 Car Attached", "3 Car Detached").

garage_list = []


for item in unique_parking_features:
    lower_item = item.lower()
    if any(keyword in lower_item for keyword in garage_keywords):
        garage_list.append(item.lower())

garage_list.remove('no garage')

print("Garage-related items:", garage_list)

Garage-related items: ['integrated garage', 'workshop in garage', 'in garage electric vehicle charging station(s)', 'carport; single', 'detached garage', 'garage faces rear', 'carport', 'single garage attached', '6 car detached', 'garage/laneway suite', 'triple garage detached', '2 car detached', 'carport; multiple', '3 car detached', '6 car attached', 'quad or more detached', '4 car detached', 'garage; underground', 'garage underbuilding', 'rv garage', 'converted garage', '1 car detached', 'golf cart garage', 'garage; single', 'garage double', 'double carport', 'carport double', 'tandem parking', '3 car attached', 'single carport', 'double garage attached', 'garage quad+', 'carport & garage', 'single garage detached', 'quad or more attached', 'double garage detached', '2 car attached', 'attached', 'garage faces side', 'attached garage', 'garage door opener', 'heated garage', '4 car attached', 'grge/double tandem', 'garage triple', 'garage', 'carport quad+', '1 car attached', '5 car de

In [27]:
def check_keywords(features, keywords):

    try:
        cleaned_features = features.replace("[", "").replace("]", "").replace('"', '').replace("'", "").replace(";", ",")
        list_to_check = [f.strip().lower() for f in cleaned_features.split(',')]
        common = set(list_to_check) & set(keywords)
        
        if common: 
            return 'Yes'  
            
    except AttributeError:
        return 'No'
    
    return 'No'

The function 'check_keywords' will check if column 'Features' and 'Parking' contain at least one pfrase from the list 'garage_list'. In this case, the nan value in column 'Garage' will be changes to "Yes".

In [28]:
df['Garage new'] = df['Garage']
df.loc[df.Garage.isna(), 'Garage new'] = df.loc[df.Garage.isna(), 'Features'].apply(lambda x: check_keywords(x, garage_list))
df['Garage new'].value_counts()

Garage new
Yes    55223
No     42588
Name: count, dtype: int64

Some rows can have information about garage in Parking column

In [29]:
df.Parking.unique()

array(['Underground, Heated Garage',
       'Attached Garage, RV, See Remarks, Heated Garage', nan, ...,
       '["No Garage", "Covered"]',
       '["1 Car Detached", "Garage/Laneway Suite"]',
       '["2 Car Attached", "Parking Pad", "RV Parking"]'], dtype=object)

In [30]:
df.loc[df.Garage == 'No', 'Garage new'] = df.loc[df.Garage == 'No', 'Parking'].apply(lambda x: check_keywords(x, garage_list))
df['Garage new'].value_counts()

Garage new
Yes    55223
No     42588
Name: count, dtype: int64

The number of not-null values does not changed. So all possible values were filled from 'Features' column.

## Let's do the same for column 'Parking'

In [31]:
df.Parking.value_counts()

Parking
[]                                                 11111
Attached Garage                                     5560
Underground                                         4924
["Attached Garage"]                                 3793
Garage                                              2566
                                                   ...  
Driveway, Open, RV Access/Parking                      1
Additional Parking, Carport                            1
Additional Parking, Carport, Driveway, Open            1
Carport; Multiple, DetachedGrge/Carport, Open          1
["2 Car Attached", "Parking Pad", "RV Parking"]        1
Name: count, Length: 4907, dtype: int64

In [32]:
parking_list = ["2 Outdoor Stalls", "Add. Parking Avail.", "Additional Parking", "Assigned",
                "Carport", "Carport & Garage", "Carport Double", "Carport Quad+", "Carport Triple", 
                "Carport; Multiple", "Carport; Single", "Multiple Driveways", "Parkade", "Parking Lot", 
                "Parking Pad", "Parking Pad Cement/Paved", "Parking Space(s)", "Parking Spaces", 
                "RV", "RV Access/Parking", "RV Gated", "RV Hookup", "RV Parking", "RV Parking Avail.", 
                "Shared Driveway", "Stall", "Tandem Parking", "Underground", "Underground Parking", 
                "Visitor Parking"
                ]

parking_list = [f.lower() for f in parking_list]

We'll replace values in a Parking column to 'Yes' and "No'

In [33]:
# first, copy the column
df['Parking new'] = df['Parking']

# if the column 'Parking' has one of the words from the parking_list, it will be replaced to 'Yes'
df['Parking new'] = df['Parking new'].apply(lambda x: check_keywords(x, parking_list))

# if the column 'Features' has one of the words from the parking_list, it will be replaced to 'Yes' in column 'Parking'
df.loc[df['Parking new'] == 'No', 'Parking new'] = df.loc[df['Parking new']  == 'No', 'Features'].apply(lambda x: check_keywords(x, parking_list))

df['Parking new'].value_counts()

Parking new
No     67834
Yes    29977
Name: count, dtype: int64

## Basement

In [34]:
df.Basement.value_counts()

Basement
[]                                                      5502
["Finished", "Full"]                                    3293
Finished, Full                                          3194
Finished, N/A                                           2178
["Unfinished", "Full"]                                  1481
                                                        ... 
["Crawl Space", "Finished", "Walk-Out Access"]             1
Not Full Height, Partially Finished, Walk-Out Access       1
["Partial", "Unfinished", "Walk-Out Access"]               1
["With Windows"]                                           1
["Cellar", "Partial Basement", "Unfinished"]               1
Name: count, Length: 636, dtype: int64

Let's extract unique features and select related to 'Basement' values

In [35]:
unique_basement_features = get_unique_values('Basement')
print(len(unique_basement_features), unique_basement_features)

38 {'', 'Apartment in basement', 'Walkout', 'Suite', 'Partial Basement', 'Partial', 'Not Applicable', 'Dugout', 'Partially Finished', 'Partially finished', 'Other', 'Remodeled', 'Walk-Out To Grade', 'Not applicable', 'Slab', 'Walk-up', 'Unfinished', 'Crawl', 'Cellar', 'N/A', 'With Windows', 'Full', 'Walk out', 'Separate/Exterior Entry', 'Unknown', 'Walk-Out Access', 'Crawl Space', 'Fully Finished', 'Finished', 'Crawl space', 'None', 'No Basement', 'Walk-Up To Grade', 'Remodeled Basement', 'Not Full Height', 'Full Basement', 'Separate entrance', 'See Remarks'}


#### Finished (These basements or spaces are completed or usable, often with features like exterior access or fully finished rooms.)
- "Full Basement": A basement that spans the entire footprint of the house and is fully developed for use.
- "Full": Another way to refer to a completely finished basement.
- "Dugout": A basement or below-ground space that has been excavated (dug out) to provide additional space.
- "Fully Finished": A basement that is 100% completed with flooring, walls, ceiling, and sometimes utilities.
- "Finished": A general term for a basement that is move-in ready.
- "Remodeled Basement": A basement that was previously unfinished or outdated but has been renovated.
- "Apartment in Basement": A basement that has been converted into a separate living unit with amenities like a kitchen and bathroom.
- "Suite": A finished basement space designed as a separate living area, often with its own entrance.
- "Walk-Out Access": A basement with a door leading directly outside, typically to ground level.
- "Walk-Out To Grade": Similar to "Walk-Out Access," meaning the basement exits directly to ground level.
- "Walk-Up To Grade": A basement that has a stairway leading up to ground level for external access.
- "Walk-Out / Walkout": Another way to describe a basement that has an external door leading outside.
- "Walk-Up": A basement with stairs leading to an outside entrance.
- "With Windows": A basement that has windows, often allowing natural light and making it more livable.
- "Separate/Exterior Entry": A basement that has a private entrance from outside, making it ideal for rental units or guest spaces.
- "Separate Entrance": Similar to "Separate/Exterior Entry," meaning the basement has independent access from outside.

#### Partial (These basements or spaces are partially finished or incomplete.)
- "Partially Finished" / "Partially finished": Some areas are finished, but others may be unfinished or under construction.
- "Partial" / "Partial Basement": The basement is smaller than the home's full footprint or has limited finished space.
- "Not Full Height": A basement that does not have standard ceiling height, making it less functional as a full living space.
- "Cellar": An older term, often referring to basements used for storage, with lower ceilings and minimal finishing.

#### No Basement (These properties either don't have basements or have limited, unusable spaces underneath.)
- "Unfinished": A basement that exists but has no completed flooring, walls, or ceiling.
- "No Basement": The property does not include a basement.
- "Slab": The house is built directly on a concrete slab, with no basement underneath.
- "Crawl Space" / "Crawl": A shallow space under the house, not intended for living but used for ventilation or storage.
- "None": No basement or similar structure exists in the property.
- "N/A": Not available or not provided.
- "Not Applicable": The basement condition does not apply, possibly because there is no basement or it's an irrelevant listing.

In [36]:
basement_types = {'Finished': ['full basement', 'full', 'dugout', 'fully finished', 'finished', 'remodeled basement', 
                               'apartment in basement', 'suite', 'walk-out access', 'walk-out to grade', 'walk-up to grade', 
                               'walk-out', 'walkout', 'walk-up', 'with windows', 'separate/exterior entry', 'separate entrance'], 
                  'Partial': ['partially finished', 'partial', 'partially finished', 'partial basement', 
                              'not full height', 'cellar'], 
                  'No basement': ['unfinished', 'no basement', 'slab', 'crawl space', 'crawl', 'none', 
                                  'not applicable', 'n/a']}

In [37]:
def check_type(features, types):

    try:
        cleaned_features = features.replace("[", "").replace("]", "").replace('"', '').replace("'", "").replace(";", ",")
        list_to_check = [f.strip().lower() for f in cleaned_features.split(',')]

        for basement_type, basement_type_list in types.items():
            if set(list_to_check) & set(basement_type_list): 
                return basement_type
        return np.nan

    except AttributeError:
        return np.nan

In [38]:
df['Basement new'] = df['Basement']
df['Basement new'] = df['Basement new'].apply(lambda x: check_type(x, basement_types))
df['Basement new'].value_counts()

Basement new
Finished       30106
No basement     5471
Partial         2055
Name: count, dtype: int64

## Exterior

In [39]:
df.Exterior.value_counts().head()

Exterior
[]           9829
Brick        7340
["Brick"]    3220
["Vinyl"]    3017
Concrete     2920
Name: count, dtype: int64

In [40]:
unique_exterior_features = get_unique_values('Exterior')   
print(len(unique_exterior_features), unique_exterior_features)

140 {'', 'Stone', 'Insul Brick', 'Cul-De-Sac', 'Partially Fenced', 'Log', 'View City', 'Hardboard', 'Rock Imitation', 'Wood siding', 'Public Swimming Pool', 'Wood shingles', 'Outdoor Grill', 'Garden', 'Sloping Lot', 'Tennis Court(s)', 'River Valley View', 'Waterfront Property', 'Gated Community', 'BBQ gas line', 'Brick Veneer', 'Private Park Access', 'Flat Site', 'Private Yard', 'Outdoor Kitchen', 'Golf Nearby', 'Low Maintenance Yard', 'Back Lane', 'Boating', 'View Downtown', 'Misting System', 'River View', 'Uncovered Courtyard', 'Rain Gutters', 'Siding', 'Partially Landscaped', 'Other', 'Wood Siding', 'Balcony', 'Awning(s)', 'Vinyl', 'Dock', 'Boat Slip', 'Fencing: Partial', 'View Lake', 'Vinyl siding', 'Masonite', 'Basketball Court', 'Cedar shingles', 'Lighting', 'Aluminum/Vinyl', 'Concrete', 'Not Landscaped', 'Outdoor Shower', 'Gas Grill', 'Fenced', 'Creek', 'No Back Lane', 'Playground Nearby', 'Park/Reserve', 'Backs Onto Park/Trees', 'Beach Access', 'Water Feature', 'Playground', 'A

In [41]:
exterior_materials = {
    'Metal': ['aluminum', 'aluminum siding', 'aluminum/vinyl', 'colour loc', 'metal', 'steel'],
    'Brick': ['brick', 'brick facing', 'brick imitation', 'brick veneer'],
    'Concrete': ['concrete', 'concrete siding', 'concrete/stucco', 'concrete block', 'insul brick', 'stucco'],
    'Wood': ['cedar', 'cedar shingles', 'cedar siding', 'wood', 'wood siding', 'wood shingles', 'wood shingles', 'wood siding'],
    'Composite': ['composite siding', 'hardboard', 'masonite', 'shingles'],
    'Vinyl': ['vinyl', 'vinyl siding', 'asbestos', 'siding']
}

In [42]:
df['Exterior new'] = df['Exterior']
df['Exterior new'] = df['Exterior new'].apply(lambda x: check_type(x, exterior_materials))
df['Exterior new'].value_counts()

Exterior new
Brick        20344
Vinyl        10806
Concrete      7639
Metal         3561
Wood          2940
Composite      784
Name: count, dtype: int64

## Fireplace

In [43]:
print(df.Fireplace.shape[0]) 
df.Fireplace.value_counts()[:10]

97811


Fireplace
[]         21932
Yes        19967
1           3291
["1"]       2717
0           1454
["2"]       1256
["0"]       1171
2           1102
Gas         1014
["Gas"]      894
Name: count, dtype: int64

Column 'Fireplace' contain information in different appearence. Let's reduce it ot 'Yes' and 'No'. Also we need to get information from the column 'Fireplace Features'.

In [44]:
df['Fireplace new'] = df['Fireplace']

df.loc[df['Fireplace new'].isin(['0', '[]','["0"]']), 'Fireplace new'] = np.nan
df.loc[df['Fireplace new'].notna(), 'Fireplace new'] = 'Yes'
df.loc[~df['Fireplace new'].notna(), 'Fireplace new'] = 'No'


In [45]:
df['Fireplace new'].value_counts()

Fireplace new
No     63356
Yes    34455
Name: count, dtype: int64

# Heating

In [46]:
df.Heating.unique().shape

(2642,)

In [47]:
unique_heating_features = get_unique_values('Heating')
print(len(unique_heating_features), unique_heating_features)

85 {'', 'Fireplace(s)', 'Outside Furnace', 'Electric baseboard units', 'Geothermal', 'Sep. HVAC Units', 'Exhaust Fan', 'Hot Water', 'Fireplace Insert', 'Pellet Stove', 'Wall Furnace', 'Radiator', 'Radiant/Infra-red Heat', 'See remarks', 'Solar', 'No heat', 'High-Efficiency Furnace', 'Floor Model Furnace', 'Combination', 'ENERGY STAR Qualified Equipment', 'Furnace', 'Other', 'Space Heaters', 'Air Conditioning', 'Make-up Air', 'Gravity Heat System', 'Forced Air-1', 'Central Heat Pump', 'Radiant heat', 'In Floor', 'Floor heat', 'Geo Thermal', 'Radiant Heat', 'Standard', 'Hot water radiator heat', 'Pellet', 'Electric', 'Fan Coil', 'Heat Recovery', 'Baseboard', 'Ceiling', 'Humidity Control', 'Wood', 'Space Heater', 'Overhead Unit Heater', 'High Efficiency', 'Natural Gas', 'Forced Air-2', 'Heat Recovery Ventilation (HRV)', 'Propane', 'Heat Pump', 'Radiant Floor', 'Gravity', 'In Floor Roughed-In', 'Oil', 'Central', 'Steam', 'Forced air', 'Propane Gas', 'Natural gas', 'Separate Water Heaters',

Let's divide them in categories:

In [48]:
heating_categories = {
    'forced air': ['forced air', 'forced air-1', 'forced air-2', 'furnace'],
    'boiler': ['boiler', 'hot water', 'hot water radiator heat', 'steam', 'steam radiator'],
    'heat pump': ['central heat pump', 'heat pump', 'wall mounted heat pump'],
    'radiant': ['radiant', 'radiant heat', 'radiant ceiling', 'radiant floor', 
                'radiant/infra-red heat', 'baseboard', 'baseboard heaters', 'electric baseboard units'],
    'fireplace': ['fireplace(s)', 'fireplace insert', 'wood stove', 'pellet stove', 'coal stove', 'stove'],
    'space heat': ['space heater', 'space heaters', 'wall furnace', 'floor furnace', 
                   'floor model furnace', 'overhead heaters', 'overhead unit heater', 'ductless'],
    'alt heat': ['geo thermal', 'geothermal', 'solar', 'gravity', 'gravity heat system', 
                 'oil', 'propane', 'propane gas', 'coal'],
    'no heat': ['no heat', 'none'],  # only contains no heat options
    'other': ['mixed', 'combination', 'sep. hvac units'],  # these suggest non-specific or mixed systems
}

In [49]:
df['Heating new'] = df['Heating']
df['Heating new'] = df['Heating new'].apply(lambda x: check_type(x, heating_categories))

df['Heating new'] .value_counts()

Heating new
forced air    53347
radiant       12053
heat pump      5606
boiler         3924
alt heat        529
fireplace       238
no heat          76
space heat       74
other            64
Name: count, dtype: int64

# Flooring

In [50]:
df.Flooring.value_counts()

Flooring
[]                                                                   21368
Laminate                                                              2292
Hardwood                                                              1905
["Laminate"]                                                          1225
["Hardwood"]                                                           719
                                                                     ...  
["Hardwood", "Laminate", "Carpeted", "Other", "Wood"]                    1
["Hardwood", "Laminate", "Carpeted", "Linoleum", "Vinyl", "Cork"]        1
["Hardwood", "Laminate", "Ceramic Tile", "Cork"]                         1
["Tile", "Carpeted", "Ceramic", "Linoleum", "Wood"]                      1
["Laminate", "Tile", "Vinyl", "Wood"]                                    1
Name: count, Length: 2105, dtype: int64

In [51]:
unique_flooring_features = get_unique_values('Flooring')
print(len(unique_flooring_features), unique_flooring_features)

42 {'Bamboo', 'Stone', 'Cork Flooring', 'Cork', 'Carpeted', 'Laminate Flooring', 'Wood', 'Vinyl Plank', 'Subfloor', '', 'Ceramic/Porcelain', 'Other', 'Carpet', 'Carpet Over Softwood', 'Porcelain Tile', 'Ceramic', 'Vinyl', 'Marble', 'Granite', 'Basement Sub-Floor', 'Engineered hardwood', 'Heavy loading', 'Wall-to-wall carpet', 'Wall to Wall Carpet', 'Engineered Wood', 'Mixed', 'Non-Ceramic Tile', 'Concrete', 'Hardwood', 'Concrete Slab', 'Slate', 'Linoleum', 'Basement Slab', 'Mixed Flooring', 'Laminate', 'Ceramic Tile', 'Softwood', 'Tile', 'Cushion/Lino/Vinyl', 'Carpet over Hardwood', 'See Remarks', 'Parquet'}


Let's divide into next categories:
- Carpet: Includes all variations of carpet.
- Wood: Includes bamboo, engineered wood, hardwood, etc.
- Tile: Covers ceramic, porcelain, slate, and other types of tiles.
- Vinyl: Includes all vinyl-based flooring types.
- Laminate: Groups laminate and laminate flooring.
- Concrete: Includes concrete and concrete slabs.
- Other: For flooring types that are not easily categorized (like marble, granite, subfloor, etc.).

In [52]:
flooring_categories = {
    'carpet': ['carpet', 'carpet over softwood', 'carpet over hardwood', 'carpeted', 'wall to wall carpet', 
               'wall-to-wall carpet'],
    'wood': ['bamboo', 'engineered wood', 'engineered hardwood', 'hardwood', 'parquet', 'softwood', 'wood'],
    'tile': ['ceramic', 'ceramic tile', 'ceramic/porcelain', 'porcelain tile', 'non-ceramic tile', 'slate', 
             'stone', 'tile'],
    'vinyl': ['cushion/lino/vinyl', 'vinyl', 'vinyl plank'],
    'laminate': ['laminate', 'laminate flooring'],
    'concrete': ['concrete', 'concrete slab'],
    'other': ['basement slab', 'basement sub-floor', 'granite', 'heavy loading', 'linoleum', 'marble', 'mixed', 
              'mixed flooring', 'see remarks', 'subfloor', 'other']
}

In [53]:
df['Flooring new'] = df['Flooring']
df['Flooring new'] = df['Flooring new'].apply(lambda x: check_type(x, flooring_categories))

df['Flooring new'].value_counts()

Flooring new
carpet      12529
wood        10973
laminate     4223
tile         4071
vinyl        2197
other        1009
concrete       86
Name: count, dtype: int64

# Roof

In [54]:
df.Roof.value_counts()

Roof
[]                                          23661
Asphalt Shingles                             3534
["Asphalt Shingle"]                          2916
["Asphalt Shingles"]                         2907
Asphalt Shingle                              2486
                                            ...  
["Vinyl", "Membrane", "Mixed", "Rubber"]        1
Flat, Rubber                                    1
Flat Torch Membrane, Shingle                    1
["Asphalt", "Vinyl", "Wood"]                    1
["Asphalt Shingles", "Tar & Gravel"]            1
Name: count, Length: 247, dtype: int64

In [55]:
unique_roof_features = get_unique_values('Roof')
print(len(unique_roof_features), unique_roof_features)

51 {'', 'SBS Roofing System', 'Cedar shake', 'Fiberglass Shingles', 'Wood', 'Rubber', 'Rolled/Hot Mop', 'Flat', 'Asphalt Shingles', 'Other', 'Asphalt', 'Tar/Gravel', 'Vinyl', 'Tar &amp; Gravel', 'Asphalt Shingle', 'Shingle', 'Clay Tile', 'Pine Shakes', 'Asphalt Torch On', 'EPDM Membrane', 'Concrete Tiles', 'Wood Shingle', 'Asphalt Rolled', 'Asphalt shingle', 'Asphalt/Gravel', 'Metal Shingles', 'Unknown', 'Conventional', 'Fibreglass Shingle', 'Mixed', 'Pine Shake', 'Membrane', 'Vinyl Shingles', 'Flat Torch Membrane', 'Concrete', 'Fiberglass', 'Slate', 'Tin', 'Metal', 'Tar & gravel', 'Cedar Shakes', 'Shake', 'Cedar Shake', 'Roll Roofing', 'None', 'Wood Shingles', 'Asphalt & Gravel', 'Steel', 'Tar & Gravel', 'Tile', 'See Remarks'}


Categories:
- Asphalt: Includes all types of asphalt and asphalt shingles.
- Cedar Shake: Includes cedar shake and shakes.
- Clay: For clay tiles.
- Concrete: Covers concrete and concrete tiles.
- Fiberglass: Includes fiberglass and fiberglass shingles.
- Flat: Includes flat roofing, EPDM, membrane, and flat torch membrane types.
- Metal: Covers metal, metal shingles, steel, and tin.
- Pine Shake: Includes pine shake and shakes.
- Rubber: For rubber roofing.
- SBS: For SBS roofing system.
- Shake: Covers shake roofing.
- Shingle: Includes regular shingles and vinyl shingles.
- Slate: For slate roofing.
- Tar: Includes various types of tar and gravel roofing.
- Tile: For tile roofing.
- Wood: Covers wood shingles and wood shake.
- Other: For unconventional or mixed roofing, or cases where the material is unclear (like 'conventional', 'other', 'mixed').

In [56]:
roofing_categories = {
    'asphalt': ['asphalt', 'asphalt & gravel', 'asphalt rolled', 'asphalt shingle', 'asphalt shingles', 'asphalt torch on', 'asphalt shingle', 'asphalt/gravel'],
    'cedar shake': ['cedar shake', 'cedar shakes'],
    'clay': ['clay tile'],
    'concrete': ['concrete', 'concrete tiles'],
    'fiberglass': ['fiberglass', 'fiberglass shingles', 'fibreglass shingle'],
    'flat': ['flat', 'flat torch membrane', 'membrane', 'epdm membrane'],
    'metal': ['metal', 'metal shingles', 'steel', 'tin'],
    'pine shake': ['pine shake', 'pine shakes'],
    'rubber': ['rubber'],
    'sbs': ['sbs roofing system'],
    'shake': ['shake'],
    'shingle': ['shingle', 'vinyl shingles'],
    'slate': ['slate'],
    'tar': ['tar & gravel', 'tar & gravel', 'tar &amp; gravel', 'tar/gravel'],
    'tile': ['tile'],
    'wood': ['wood', 'wood shingle', 'wood shingles'],
    'other': ['conventional', 'mixed', 'other'],
}

In [57]:
df['Roof new'] = df['Roof']
df['Roof new'] = df['Roof new'].apply(lambda x: check_type(x, roofing_categories))
df['Roof new'].value_counts()

Roof new
asphalt        14182
metal            960
fiberglass       524
other            136
flat             115
cedar shake       98
shingle           83
tile              53
shake             42
clay              40
tar               37
wood              25
rubber            20
concrete          18
pine shake        10
sbs                3
slate              1
Name: count, dtype: int64

# Property Type

In [58]:
df['Property Type'].value_counts()

Property Type
Single Family        51072
Condo                24990
Vacant Land          11268
Condo/Townhome        2914
Townhome              2727
MultiFamily           1821
Duplex                1817
Agricultural           441
Farm                   276
Mobile Home            271
Manufactured Home      213
Timeshare                1
Name: count, dtype: int64

# Waterfront

In [59]:
df.Waterfront.value_counts()

Waterfront
Yes    178
Name: count, dtype: int64

In [60]:
df['Waterfront new'] = df['Waterfront']
df.loc[df['Waterfront new'].isna(), 'Waterfront new'] = 'No'

df['Waterfront new'].value_counts()

Waterfront new
No     97633
Yes      178
Name: count, dtype: int64

# Subdivision

In [61]:
df.Subdivision.value_counts()[:30]

Subdivision
NONE                    1455
Whalley                  718
Willoughby Heights       511
Downtown VW              479
Coquitlam West           374
Brighouse                367
Brentwood Park           351
Metrotown                311
Abbotsford West          299
White Rock               285
Central Abbotsford       283
Fleetwood Tynehead       279
Mission BC               271
Beltline                 270
Langley City             242
Yaletown                 241
West Cambie              210
West End VW              210
Cloverdale BC            202
King George Corridor     201
Abbotsford East          197
Sullivan Station         194
West Newton              194
Grandview Surrey         193
Cambie                   184
Collingwood VE           183
East Newton              182
Marpole                  179
Guildford                165
West Central             157
Name: count, dtype: int64

# Sewer

In [62]:
df.Sewer.value_counts()

Sewer
[]                                                        10549
Sanitary sewer                                             9671
["Municipal sewage system"]                                7379
["Sanitary sewer"]                                         4812
["Septic System"]                                          4682
                                                          ...  
Perc Test On File, Septic Needed                              1
["Perc Test On File", "Septic Needed"]                        1
["Sewer Available", "Sewer Connected", "Sewer To Lot"]        1
Septic System: Common, Sewer To Lot                           1
["Municipal/Community", "Septic Tank &amp; Field"]            1
Name: count, Length: 209, dtype: int64

In [63]:
unique_sewer_features = get_unique_values('Sewer')
print(len(unique_sewer_features), unique_sewer_features)

45 {'', 'Septic Field', 'Septic Tank &amp; Field', 'Outhouse', 'Mound Septic', 'Other', 'Storm sewer', 'Septic Needed', 'Private Sewer', 'Aerobic Treatment System', 'Open Discharge', 'Private sewer', 'Mound', 'Septic System', 'Shared Septic', 'Sewer', 'Sewer Connected', 'Septic tank', 'Liquid Surface Dis', 'Gravity', 'Lagoon', 'No sewage system', 'Septic Tank &amp; Mound', 'Pump', 'Sanitary sewer', 'Unknown', 'Public Sewer', 'Holding Tank', 'Tank &amp; Straight Discharge', 'Engineered Septic', 'Sewer Available', 'Low Pressure Sewage Sys', 'Septic Tank and Field', 'Sewer To Lot', 'See remarks', 'Outflow Tank', 'Septic System: Common', 'Attached to Municipal', 'None', 'Municipal/Community', 'Perc Test On File', 'Municipal sewage system', 'Septic Tank', 'See Remarks', 'Facultative Lagoon'}


In [64]:
sewage_categories = {
    'municipal': ['municipal/community', 'municipal sewage system', 'sanitary sewer', 'sewer', 'sewer connected', 
                  'sewer to lot', 'sewer available', 'public sewer', 'attached to municipal'],
    'septic': ['septic tank', 'septic system', 'septic system: common', 'septic field', 'septic tank and field', 
               'septic tank & mound', 'mound septic', 'septic tank & field', 'septic needed', 'engineered septic'],
    'private': ['private sewer', 'private sewer', 'holding tank', 'low pressure sewage sys', 'shared septic'],
    'alternative': ['aerobic treatment system', 'facultative lagoon', 'lagoon', 'outflow tank', 'open discharge', 
                    'liquid surface dis', 'pump', 'tank & straight discharge'],
    'none': ['no sewage system', 'outhouse', 'none'],
}

In [65]:
df['Sewer new'] = df['Sewer']
df['Sewer new'] = df['Sewer new'].apply(lambda x: check_type(x, sewage_categories))
df['Sewer new'] = df['Sewer new'].fillna('none')

df['Sewer new'].value_counts()

Sewer new
none           60523
municipal      28405
septic          8447
private          273
alternative      163
Name: count, dtype: int64

# Additional Features 

Let's add some additional features like Pool, Gerden, View and Balcony. 

In [66]:
mix_features = ['Basement', 'Exterior', 'Features', 'Fireplace', 'Garage', 'Heating', 'Parking']

In [67]:
df['Combined'] = df[mix_features].astype(str).agg(','.join, axis=1)
df.head(1)

streetAddress addressLocality  \
7  1701 Coursier Avenue Unit# 1301, Revelstoke, BC V0E2S3      Revelstoke   

  addressRegion postalCode   latitude   longitude  \
7            BC     V0E2S3  50.976585 -118.173149   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     description  \
7  Don't miss out on this rare 3 bedroom, 2 bath corner unit at Mackenzie Plaza! This fully furnished turn-key short term rental with proven revenue is ready for its new owner. Sit on the patio and enjoy the evening sun of this west facing condo which overlooks Mt Macpherson, Boulder Mountain and Frisbee Ridge. These units can be rented short term, long term or lived in permanently. Included is an underground parking stall and storage locker. The building features elevator access from parking. This amazing location is only minutes to Revelstoke Mountain Resort and downtown Revelstoke, and directly across the road from the new Cabot Revelstoke golf course. Take advantage of the commercial amenities that await on the ground floor which includes a grocery store, co-working space, gym, liquor store, art gallery and more. The resort shuttle bus stops directly across the street so leave your car at home. Book a showing today! (id:38686)   

      price priceCurrency property-beds property-baths property-sqft  Acreage  \
7  839000.0           CAD           3.0            2.0           891      0.0   

  Air Conditioning Basement Bath Exterior                    Features  \
7              NaN      NaN  0.0      NaN  Underground, Heated Garage   

  Fireplace Garage    Heating    MLS® # Property Tax Property Type Roof  \
7       NaN    NaN  Heat Pump  10334677       $3,478         Condo  NaN   

                     Sewer Square Footage Subdivision Waterfront  \
7  Municipal sewage system       891 SQFT         NaN        NaN   

                      Parking               Flooring Fireplace Features  \
7  Underground, Heated Garage  Tile, Carpeted, Vinyl                NaN   

  Square Footage new Garage new Parking new Basement new Exterior new  \
7                891        Yes         Yes          NaN          NaN   

  Fireplace new Heating new Flooring new Roof new Waterfront new  Sewer new  \
7            No   heat pump       carpet      NaN             No  municipal   

                                                                          Combined  
7  nan,nan,Underground, Heated Garage,nan,nan,Heat Pump,Underground, Heated Garage

In [68]:
##### Extract unique features
unique_mix_features = set()

for features in df['Combined'].dropna():
    # Remove brackets and extra quotes
    cleaned_features = features.replace("[", "").replace("]", "").replace('"', '').replace("'", "")
    # Split by comma
    feature_list = [f.strip() for f in cleaned_features.split(",")]
    unique_mix_features.update(feature_list)

print(len(unique_mix_features), unique_mix_features)

457 {'Primary Bedroom', 'Wood Burning', '', 'Gas', 'Fireplace(s)', 'Partial Basement', 'Log', 'Titled', 'Front Drive Access', 'Wood siding', '7', 'Public Swimming Pool', 'Glass Doors', 'Bedroom', 'Crawl', 'Integrated Garage', 'Outdoor Grill', 'Garden', 'Brass', 'Waterfront Property', 'Underground Parking', 'Unknown', 'In Garage Electric Vehicle Charging Station(s)', 'Gated Community', 'BBQ gas line', 'nan', 'Interlocked', 'Flat Site', 'Rear Drive Access', 'High-Efficiency Furnace', 'Private Yard', 'Heatilator', 'Golf Nearby', 'Boating', 'Offset', 'River View', 'Uncovered Courtyard', 'Stall', '6 Car Detached', 'Furnace', 'Other', 'Yes', 'Dock', 'Air Conditioning', 'Loft', 'Gravity Heat System', 'Radiant heat', 'Floor heat', 'EPA Certified Wood Stove', 'Gravel', 'Geo Thermal', 'Cedar shingles', 'Parking Space(s)', 'Outdoor Shower', 'Fenced', 'No Back Lane', 'Enclosed', 'Indoor', 'Aluminum siding', 'Courtyard', 'Airport Nearby', 'Fan Coil', 'Heat Recovery', 'Ceiling', 'Drive Through', 'St

# Pool

Let's add Pool feature

In [69]:
pool_features = ["swimming pool", "public swimming pool"]

In [70]:
df['Pool'] = df['Combined'].apply(lambda x: check_keywords(x, pool_features))
df['Pool'].value_counts()

Pool
No     96992
Yes      819
Name: count, dtype: int64

# Garden

In [71]:
garden_features = ["vegetable garden", "garden", "fruit trees/shrubs", "private yard", 
                   "partially landscaped", "landscaped"]

In [72]:
df['Garden'] = df['Combined'].apply(lambda x: check_keywords(x, garden_features))
df['Garden'].value_counts()

Garden
No     92391
Yes     5420
Name: count, dtype: int64

# View

In [73]:
view_features = ["View Downtown", "River View", "View City", "View Lake", "Lake View", 
                 "Ravine View", "River Valley View"]

view_features = [f.lower() for f in view_features]

In [74]:
def check_view(features, keywords):

    try:
        cleaned_features = features.replace("[", "").replace("]", "").replace('"', '').replace("'", "").replace(";", ",")
        list_to_check = [f.strip().lower() for f in cleaned_features.split(',')]
        common = set(list_to_check) & set(keywords)
        
        if common: 
            return list(common)[0]  
            
    except AttributeError:
        return np.nan
    
    return np.nan

In [75]:
df['View'] = df['Combined'].apply(lambda x: check_view(x, view_features))
df['View'].value_counts()

View
view downtown        215
river valley view    167
ravine view          126
view city             98
view lake             83
lake view             58
river view            35
Name: count, dtype: int64

In [76]:
df.loc[df['View'].isin(['view lake', 'lake view']), 'View'] = 'Lake'
df.loc[df['View'].isin(['view downtown']), 'View'] = 'Downtown'
df.loc[df['View'].isin(['view city']), 'View'] = 'City'
df.loc[df['View'].isin(['river view']), 'View'] = 'River'
df.loc[df['View'].isin(['ravine view', 'river valley view']), 'View'] = 'Valley'

df['View'].value_counts()

View
Valley      293
Downtown    215
Lake        141
City         98
River        35
Name: count, dtype: int64

# Balcony

In [77]:
balcony_features = ['Balcony', 'Balcony/Deck', 'Balcony/Patio']
balcony_features = [f.lower() for f in balcony_features]

In [78]:
df['Balcony'] = df['Combined'].apply(lambda x: check_keywords(x, balcony_features))
df['Balcony'].value_counts()

Balcony
No     94342
Yes     3469
Name: count, dtype: int64

In [79]:
df = df.drop(columns=['streetAddress', 'postalCode', 'description', 'priceCurrency', 'Air Conditioning', 
                 'Basement', 'Exterior', 'Features', 'Fireplace', 'Garage', 'Heating', 'MLS® #', 'Roof', 
                 'Sewer', 'Waterfront', 'Parking', 'Flooring', 'Fireplace Features', 'Combined', 
                 'Subdivision', 'property-sqft', 'Square Footage', 'Bath', 'Property Tax'])

In [80]:
df = df.rename(columns={'addressLocality': 'City', 
                        'addressRegion': 'Province', 
                        'latitude': 'Latitude', 
                        'longitude': 'Longitude', 
                        'price': 'Price',
                        'property-baths': 'Bathrooms', 
                        'property-beds': 'Bedrooms', 
                        'Square Footage new': 'Square Footage', 
                        'Garage new': 'Garage',	
                        'Parking new': 'Parking',	
                        'Basement new': 'Basement', 
                        'Exterior new': 'Exterior', 
                        'Fireplace new': 'Fireplace', 
                        'Heating new': 'Heating', 
                        'Flooring new': 'Flooring', 
                        'Roof new': 'Roof', 
                        'Waterfront new': 'Waterfront', 
                        'Sewer new': 'Sewer'})

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 97811 entries, 7 to 133724
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   City            97811 non-null  object 
 1   Province        97811 non-null  object 
 2   Latitude        97811 non-null  object 
 3   Longitude       97811 non-null  object 
 4   Price           97811 non-null  object 
 5   Bedrooms        84879 non-null  object 
 6   Bathrooms       97041 non-null  object 
 7   Acreage         97811 non-null  float64
 8   Property Type   97811 non-null  object 
 9   Square Footage  46732 non-null  object 
 10  Garage          97811 non-null  object 
 11  Parking         97811 non-null  object 
 12  Basement        37632 non-null  object 
 13  Exterior        46074 non-null  object 
 14  Fireplace       97811 non-null  object 
 15  Heating         75911 non-null  object 
 16  Flooring        35088 non-null  object 
 17  Roof            16347 non-null  obj

# Types

In [82]:
df['Square Footage'] = df['Square Footage'].str.replace(',','')

In [83]:
number_columns = ['Latitude', 'Longitude', 'Price', 'Bedrooms', 'Bathrooms', 'Acreage', 
                  'Square Footage']

for col in number_columns: 
    df[col] = df[col].astype(float)

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 97811 entries, 7 to 133724
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   City            97811 non-null  object 
 1   Province        97811 non-null  object 
 2   Latitude        97811 non-null  float64
 3   Longitude       97811 non-null  float64
 4   Price           97811 non-null  float64
 5   Bedrooms        84879 non-null  float64
 6   Bathrooms       97041 non-null  float64
 7   Acreage         97811 non-null  float64
 8   Property Type   97811 non-null  object 
 9   Square Footage  46732 non-null  float64
 10  Garage          97811 non-null  object 
 11  Parking         97811 non-null  object 
 12  Basement        37632 non-null  object 
 13  Exterior        46074 non-null  object 
 14  Fireplace       97811 non-null  object 
 15  Heating         75911 non-null  object 
 16  Flooring        35088 non-null  object 
 17  Roof            16347 non-null  obj

# Remove inconsistent data

In [85]:
data = df.copy()

### Square Footage, Bedrooms, and Bathrooms

Let's remove all houses with out Bedroom or Bathrooms

In [86]:
data.dropna(subset=["Bedrooms", "Bathrooms", "Square Footage"], inplace=True)
data[["Bedrooms", "Bathrooms", "Square Footage"]].describe()

Bedrooms     Bathrooms  Square Footage
count  45188.000000  45188.000000    45188.000000
mean       3.207245      2.513477     1780.138045
std        1.647695      1.428373     1985.084438
min        0.000000      0.000000        0.000000
25%        2.000000      2.000000      969.000000
50%        3.000000      2.000000     1410.000000
75%        4.000000      3.000000     2191.000000
max       38.000000     26.000000   290034.000000

Remove also rows with 'Square Footage' less then 120 sqft.

In [87]:
data = data[data["Square Footage"] > 120]

Remove rows with no bedrooms or bathrooms

In [88]:
# data = data[(data["Bedrooms"] > 0) & (data["Bathrooms"] > 0)]

In [89]:
data[["Bedrooms", "Bathrooms", "Square Footage"]].describe()

Bedrooms     Bathrooms  Square Footage
count  45086.000000  45086.000000    45086.000000
mean       3.209644      2.516524     1784.155348
std        1.636750      1.424381     1985.528762
min        0.000000      0.000000      140.000000
25%        2.000000      2.000000      972.000000
50%        3.000000      2.000000     1413.000000
75%        4.000000      3.000000     2194.000000
max       37.000000     26.000000   290034.000000

### Province and City

In [90]:
data.Province.value_counts()

Province
BC    23144
SK     5723
AB     5317
ON     3872
NS     2418
NB     2237
NL     1417
PE      799
MB       93
YT       53
NT       13
Name: count, dtype: int64

In [91]:
data.City.unique().shape

(3171,)

In [92]:
data.City.value_counts()[:10]

City
Vancouver    3861
Surrey       3454
Calgary      2176
Richmond     1377
Kelowna      1356
Burnaby      1340
Langley      1191
Saskatoon    1066
Regina        975
Coquitlam     971
Name: count, dtype: int64

We have 3112 different cities in dataset. Perhebs, we will not use this column while modeling

### Price

In [93]:
data.Price.value_counts()

Price
499900.0     385
599900.0     330
699900.0     284
399900.0     279
599000.0     277
            ... 
1858900.0      1
1085952.0      1
496000.0       1
1633000.0      1
1109000.0      1
Name: count, Length: 5538, dtype: int64

We have a lot of houses with unrealistic low price. Perhebs. some position from rent accidantly were put inro sell patr. Let's delete all houses with price less then 50 000.

In [94]:
data = data[data.Price >= 50_000]

### Property Type

In [95]:
data['Property Type'].value_counts()

Property Type
Single Family        26619
Condo                12593
Townhome              2237
Condo/Townhome        1655
Duplex                1056
MultiFamily            281
Mobile Home            228
Manufactured Home      211
Vacant Land             16
Name: count, dtype: int64

### Garage, Parking, 

In [96]:
data[['Garage', 'Parking']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 44896 entries, 7 to 133724
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Garage   44896 non-null  object
 1   Parking  44896 non-null  object
dtypes: object(2)
memory usage: 1.0+ MB


In [97]:
data['Garage'].value_counts()

Garage
Yes    31600
No     13296
Name: count, dtype: int64

In [98]:
data['Parking'].value_counts()

Parking
No     26099
Yes    18797
Name: count, dtype: int64

### Fireplace, Pool, Garden, Balcony

In [99]:
data[['Fireplace', 'Pool', 'Garden', 'Balcony', 'Sewer']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 44896 entries, 7 to 133724
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Fireplace  44896 non-null  object
 1   Pool       44896 non-null  object
 2   Garden     44896 non-null  object
 3   Balcony    44896 non-null  object
 4   Sewer      44896 non-null  object
dtypes: object(5)
memory usage: 2.1+ MB


### Basement, Exterior, Fireplace, Heating, Flooring, Roof, Sewer, View

In [100]:
data[['Basement', 'Exterior', 'Heating', 'Flooring', 'Roof', 'View']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 44896 entries, 7 to 133724
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Basement  14962 non-null  object
 1   Exterior  17439 non-null  object
 2   Heating   38242 non-null  object
 3   Flooring  15216 non-null  object
 4   Roof      10052 non-null  object
 5   View      12 non-null     object
dtypes: object(6)
memory usage: 2.4+ MB


In [101]:
# data = data.drop(columns=['View'])
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 44896 entries, 7 to 133724
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   City            44896 non-null  object 
 1   Province        44896 non-null  object 
 2   Latitude        44896 non-null  float64
 3   Longitude       44896 non-null  float64
 4   Price           44896 non-null  float64
 5   Bedrooms        44896 non-null  float64
 6   Bathrooms       44896 non-null  float64
 7   Acreage         44896 non-null  float64
 8   Property Type   44896 non-null  object 
 9   Square Footage  44896 non-null  float64
 10  Garage          44896 non-null  object 
 11  Parking         44896 non-null  object 
 12  Basement        14962 non-null  object 
 13  Exterior        17439 non-null  object 
 14  Fireplace       44896 non-null  object 
 15  Heating         38242 non-null  object 
 16  Flooring        15216 non-null  object 
 17  Roof            10052 non-null  obj

# Save data

In [104]:
data.to_csv(DATA_PATH + '/cleaned_canada.csv', index=False)
print(f"Cleaned data is saved at {DATA_PATH+'/cleaned_canada.csv'}")
# data.to_csv("/home/aircsrv5/Quan/DataOps/DataOps-bootstrap/demo/include/data/cleaned_canada.csv", index=False)